File: javascript/lesson 11/solveNQueens.js
// https://leetcode.com/problems/n-queens/submissions/

function solveNQueens(n) {
    const result = [];
    const board = [...new Array(n)].map(() => [...new Array(n)].map(() => '.'));
    
    solveNQueensImpl(board, 0, result, new Set(), new Set(), new Set());
    
    return result;
}


function solveNQueensImpl(board, row, result, columns, diagonals, antiDiagonals) {
    if (row === board.length) {
        result.push(board.map((line) => line.join("")));
        return;
    }
    
    for (let col = 0; col < board[0].length; col++) {
        const diag = row - col;
        const antiDiag = row + col;

        if (columns.has(col) || diagonals.has(diag) || antiDiagonals.has(antiDiag)) continue;

        board[row][col] = 'Q';                
        columns.add(col);
        diagonals.add(diag);
        antiDiagonals.add(antiDiag);

        solveNQueensImpl(board, row + 1, result, columns, diagonals, antiDiagonals);

        board[row][col] = '.'; // backtrack
        columns.delete(col);
        diagonals.delete(diag);
        antiDiagonals.delete(antiDiag);
    }
}

File: ruby/lesson 11/solve_n_queens.rb
# https://leetcode.com/problems/n-queens/submissions/

def solve_n_queens(n)
    result = []
    board = Array.new(n) { Array.new(n) { '.' } }
    
    solve_n_queens_impl(board, 0, result, Set.new, Set.new, Set.new)
    
    result
end

def solve_n_queens_impl(board, row, result, columns, diagonals, anti_diagonals)
    if row == board.length
        result.push(board.map { |line| line.join("") })
        return
    end
    
    (0.. board[0].length - 1).each do |col| 
        diag = row - col
        antiDiag = row + col

        next if columns.include?(col) || diagonals.include?(diag) || anti_diagonals.include?(antiDiag)

        board[row][col] = 'Q'
        columns.add(col)
        diagonals.add(diag)
        anti_diagonals.add(antiDiag)

        solve_n_queens_impl(board, row + 1, result, columns, diagonals, anti_diagonals)

        board[row][col] = '.' # backtrack
        columns.delete(col)
        diagonals.delete(diag)
        anti_diagonals.delete(antiDiag)
    end
end
